processor/deltatocumulativeprocessor/internal/delta/delta.go

```diff
@@ -0,0 +1,61 @@
+package delta
+
+import (
+	"fmt"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/streams"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+)
+
+func construct[D data.Point[D]]() streams.Aggregator[D] {
+	acc := &Accumulator[D]{dps: make(map[streams.Ident]D)}
+	lock := &Lock[D]{next: acc}
+	return lock
+}
+
+func Numbers() streams.Aggregator[data.Number] {
+	return construct[data.Number]()
+}
+
+func Histograms() streams.Aggregator[data.Histogram] {
+	return construct[data.Histogram]()
+}
+
+var _ streams.Aggregator[data.Number] = (*Accumulator[data.Number])(nil)
+
+type Accumulator[D data.Point[D]] struct {
+	dps map[streams.Ident]D
+}
+
+func (a *Accumulator[D]) Aggregate(id streams.Ident, dp D) (D, error) {
+	reset := func() {
+		a.dps[id] = dp.Clone()
+	}
+
+	aggr, ok := a.dps[id]
+	if !ok {
+		fmt.Printf("d2c: new stream %s", id)
+		reset()
+		return dp, nil
+	}
+
+	if dp.StartTimestamp() != aggr.StartTimestamp() {
```

Comment:
Could we make this behavior configurable? For my use case, it would ideal if we could aggregate delta metrics from two different sources / processes.

IE delta metric from process A: `[{t=0, 2}, {t=3, 4}, {t=4, 8}]`
and delta metric from process B: `[{t=1, 2}, {t=3, 1}, {t=6, 5}]`

Would be allowed to be combined into the single stream (cumulative): 
```[{t=0, 2}, {t=1, 4}, {t=3, 9}, {t=4, 17}, {t=6, 22}]```

NOTE: the two t=3 metrics are merged

Reply:
> aggregate delta metrics from two different sources / processes

I don't think this _necessarily_ violates the single writer principle, but it depends on whether or not these streams are differentiated from each other via labels. If the streams come from two different processes, then quite likely there _should_ be one or more labels differentiated them. In that case, aggregation is possible but is described as _spatial_ aggregation, where we aggregate multiple time series together. The data model is designed for this and I think we should support it in the collector, but in a dedicated processor e.g. `spatialaggregationprocessor`.

However, I suspect you're describing this situation:

```
In systems, there is the possibility of multiple writers sending data for the same metric stream (duplication). For example, if an SDK implementation fails to find uniquely identifying Resource attributes for a component, then all instances of that component could be reporting metrics as if they are from the same resource. In this case, metrics will be reported at inconsistent time intervals. For metrics like cumulative sums, this could cause issues where pairs of points appear to reset the cumulative sum leading to unusable metrics.

Multiple writers for a metric stream is considered an error state, or misbehaving system. Receivers SHOULD presume a single writer was intended and eliminate overlap / deduplicate.
```

My interpretation is that because this is specifically called out as incorrect behavior, we should not endeavor to fix it in this processor. However, it's notable that it states (emphasis mine) "Receivers SHOULD presume a single writer was intended **_and eliminate overlap / deduplicate_**." So while this is explicitly an error case, it's also one for which there should be a solution. I think probably it should be solved in another processor though. If you're interested in this, I think it would be valuable to attempt a design for a separate processor first.

Reply:
Your interpretation is correct. The incoming metrics *should* have unique label sets to differentiate, but I'd like to combine them to reduce cardinality. I'll work on a design issue for a `spacialaggregationprocessor`.

This thread can be closed as resolved.